### PR TITLE
[quant][pt2e] Allow users to override train/eval behavior

### PR DESCRIPTION
**Pull request by andrewor14**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119091

Summary: This commit adds a util for PT2E quantization users
to call `model.train()` and `model.eval()` without error.
Instead, these will automatically call the equivalent
`move_exported_model_to_train/eval` for the user, which only
switch behavior for special ops like dropout and batchnorm.
This enables users to onboard to the PT2E flow more easily.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_allow_exported_model_train_eval

Reviewers: jerryzh168, tugsbayasgalan, zhxchen17

Subscribers: jerryzh168, tugsbayasgalan, zhxchen17, supriyar

Differential Revision: [D53426636](https://our.internmc.facebook.com/intern/diff/D53426636)

### CODE REVIEW

File: torch/ao/quantization/pt2e/eval_utils.py [hunk 1]
```diff
@@ -137,3 +139,25 @@ def _move_exported_model_to_train(model: torch.fx.GraphModule):
...
+    Note: This does not achieve the same effect as what `model.train()` and `model.eval()`
+    does in eager models, but only provides an approximation. In particular, user code
+    branching on `training` flag will not function correctly in general because the branch
+    is already specialized at export time. Additionally, other ops beyond dropout and batchnorm
+    that have different train/eval behavior will also not be converted properly.
+    """
+    def _train(self, mode: bool = True):
+        _move_exported_model_to_train(self)
+
+    def _eval(self, mode: bool = True):
```
Comment (jerryzh168) [inline torch/ao/quantization/pt2e/eval_utils.py:158]: `eval` doesn't have a `mode` argument actually: https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#Module.eval

File: torch/ao/quantization/__init__.py [hunk 2]
```diff
@@ -14,6 +14,7 @@
 from .stubs import *  # noqa: F403
 from .pt2e.eval_utils import _move_exported_model_to_eval as move_exported_model_to_eval
 from .pt2e.eval_utils import _move_exported_model_to_train as move_exported_model_to_train
+from .pt2e.eval_utils import _allow_exported_model_train_eval_for_special_ops as allow_exported_model_train_eval_for_special_ops
```
Comment (jerryzh168) [inline torch/ao/quantization/__init__.py:17]: nit: maybe just define these as public methods in `eval_utils`? not sure what is the difference between this and defining as public methods directly

**Review by jerryzh168 — APPROVED** [review]
looks great, thanks!

File: torch/ao/quantization/pt2e/eval_utils.py [hunk 3]
```diff
@@ -137,3 +139,25 @@ def _move_exported_model_to_train(model: torch.fx.GraphModule):
     _replace_dropout(model, train_to_eval=False)
     _replace_batchnorm(model, train_to_eval=False)
     return model
+
+def _allow_exported_model_train_eval_for_special_ops(model: torch.fx.GraphModule):
```
Comment (jerryzh168) [inline torch/ao/quantization/pt2e/eval_utils.py:143]: nit: for naming, maybe just `allow_exported_model_train_eval`?

File: torch/ao/quantization/pt2e/eval_utils.py (hunk 1 above)
Comment (andrewor14) [reply torch/ao/quantization/pt2e/eval_utils.py:158]: Good cath

File: torch/ao/quantization/__init__.py (hunk 2 above)
Comment (andrewor14) [reply torch/ao/quantization/__init__.py:17]: I kind of want to keep `eval_utils` a (semi) private namespace, since the file name is kind of TBD. Don't really want users to import from there directly. Do you think that's better than exposing this under top level toq?

**Review by tugsbayasgalan — APPROVED** [review]

**Review by zhxchen17 — APPROVED** [review]

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D53426636).
